File: chains/cpp/zkvm/src/vertex.cpp
// Copyright (C) 2026, Lux Industries Inc. All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause-Eco

#include "lux/zkvm/vertex.hpp"

#include "lux/zkvm/block.hpp"
#include "lux/zkvm/vm.hpp"

namespace lux::zkvm {
namespace {

void put_be64(Bytes& out, std::uint64_t v) {
    for (int i = 7; i >= 0; --i) out.push_back(std::uint8_t(v >> (8 * i)));
}

void put_be32(Bytes& out, std::uint32_t v) {
    for (int i = 3; i >= 0; --i) out.push_back(std::uint8_t(v >> (8 * i)));
}

std::uint64_t get_be64(ByteView b, std::size_t pos) {
    std::uint64_t v = 0;
    for (std::size_t i = 0; i < 8; ++i) v = (v << 8) | b[pos + i];
    return v;
}

std::uint32_t get_be32(ByteView b, std::size_t pos) {
    std::uint32_t v = 0;
    for (std::size_t i = 0; i < 4; ++i) v = (v << 8) | b[pos + i];
    return v;
}

}  // namespace

Id Vertex::compute_id() const {
    Hasher h;
    if (vm_) h.write(view(vm_->bind()));
    h.num(vertex_height);
    h.num32(epoch);
    for (const auto& p : parents) h.write(view(p));
    for (const auto& tx : txs) {
        const Id tx_id = tx.compute_id();
        h.write(view(tx_id));
    }
    return h.sum();
}

Bytes Vertex::serialize() const {
    // height(8) | epoch(4) | parentCount(4) | parents | txCount(4) | [len(4) tx]*
    Bytes out;
    out.reserve(8 + 4 + 4 + parents.size() * 32 + 4 + txs.size() * 64);
    put_be64(out, vertex_height);
    put_be32(out, epoch);
    put_be32(out, std::uint32_t(parents.size()));
    for (const auto& p : parents) out.insert(out.end(), p.begin(), p.end());
    put_be32(out, std::uint32_t(txs.size()));
    for (const auto& tx : txs) {
        const Bytes raw = tx.marshal();
        put_be32(out, std::uint32_t(raw.size()));
        out.insert(out.end(), raw.begin(), raw.end());
    }
    return out;
}

void Vertex::finish() {
    tx_ids.clear();
    tx_ids.reserve(txs.size());
    for (const auto& tx : txs) tx_ids.push_back(tx.compute_id());
    id_ = compute_id();
    bytes_ = serialize();
}

ByteSet Vertex::nullifier_set() const {
    ByteSet s;
    for (const auto& tx : txs)
        for (const auto& n : tx.nullifiers) s.insert(n);
    return s;
}

bool Vertex::conflicts(const Vertex& other) const {
    const auto ours = nullifier_set();
    for (const auto& tx : other.txs)
        for (const auto& n : tx.nullifiers)
            if (ours.count(n)) return true;
    return false;
}

wire::Result<void> Vertex::check() {
    if (vm_ == nullptr) return std::unexpected("zkvm: vertex is not bound to a chain");

    if (txs.size() > vm_->z_config().max_tx_per_block)
        return std::unexpected(std::string(kErrInvalidBlock) + ": " +
                               std::to_string(txs.size()) + " transactions over the " +
                               std::to_string(vm_->z_config().max_tx_per_block) + " cap");

    // A vertex extends the frontier. The store keeps one tip for both shapes, so
    // a vertex that names something else moves the chain sideways.
    const Id tip = vm_->chain().tip();
    const std::uint64_t tip_height = vm_->chain().height();
    if (parents.size() != 1 || parents[0] != tip)
        return std::unexpected(std::string(kErrNotOnTip) +
                               ": vertex parents do not name the tip " + hex(tip));
    if (vertex_height != tip_height + 1)
        return std::unexpected(std::string(kErrInvalidHeight) + ": height " +
                               std::to_string(vertex_height) +
                               " does not follow the tip at " + std::to_string(tip_height));

    // Every nullifier in the vertex must be distinct. admit only sees nullifiers
    // already spent in ACCEPTED state; assembly refuses to batch conflicting
    // transactions, and a vertex that arrived on the wire is held to the same
    // rule or one shielded note is spent twice.
    ByteSet spent_here;
    for (const auto& tx : txs)
        for (const auto& n : tx.nullifiers)
            if (!spent_here.insert(n).second) return std::unexpected(kErrDuplicateNullifier);

    for (const auto& tx : txs) {
        if (auto r = vm_->admit(tx, vertex_height); !r) return r;
    }
    return {};
}

wire::Result<void> Vertex::commit() {
    if (vm_ == nullptr) return std::unexpected("zkvm: vertex is not bound to a chain");
    // A vertex is not a block — several parents, no timestamp — but it changes
    // state the same way, and this is that way.
    return vm_->chain().accept(std::static_pointer_cast<Decision>(shared_from_this()));
}

wire::Result<void> Vertex::write(store::Store& view_store) {
    (void)view_store;
    for (const auto& tx : txs) {
        for (const auto& n : tx.nullifiers) {
            if (auto r = vm_->nullifiers().mark_spent(view(n), vertex_height); !r) return r;
        }
        for (std::size_t i = 0; i < tx.outputs.size(); ++i) {
            Utxo u;
            u.tx_id = tx.id;
            u.output_index = std::uint32_t(i);
            u.commitment = tx.outputs[i].commitment;
            u.ciphertext = tx.outputs[i].encrypted_note;
            u.ephemeral_pk = tx.outputs[i].ephemeral_pubkey;
            u.height = vertex_height;
            if (auto r = vm_->utxos().add(u); !r) return r;
        }
    }
    return {};
}

void Vertex::publish() {
    status_ = Status::Accepted;
    for (const auto& tx : txs) vm_->mempool().remove(tx.id);
}

// Decided once, for the reason the block is: a vertex already accepted has
// spent its notes, and handing its transactions back would put a double spend
// into the pool this node builds from.
void Vertex::reject() {
    if (status_ != Status::Processing) return;
    status_ = Status::Rejected;
    for (const auto& tx : txs) (void)vm_->mempool().add(tx);
}

wire::Result<std::shared_ptr<Vertex>> deserialize_vertex(ByteView data, Vm& vm) {
    if (data.size() < 16) return std::unexpected(kErrInvalidBlock);
    std::size_t pos = 0;

    auto v = std::make_shared<Vertex>();
    v->bind_vm(vm);
    v->vertex_height = get_be64(data, pos);
    pos += 8;
    v->epoch = get_be32(data, pos);
    pos += 4;

    const std::uint32_t parent_count = get_be32(data, pos);
    pos += 4;
    if (std::size_t(parent_count) > (data.size() - pos) / 32)
        return std::unexpected(kErrInvalidBlock);
    v->parents.resize(parent_count);
    for (std::uint32_t i = 0; i < parent_count; ++i) {
        std::copy(data.begin() + std::ptrdiff_t(pos), data.begin() + std::ptrdiff_t(pos + 32),
                  v->parents[i].begin());
        pos += 32;
    }

    if (pos + 4 > data.size()) return std::unexpected(kErrInvalidBlock);
    const std::uint32_t tx_count = get_be32(data, pos);
    pos += 4;
    // Every transaction costs at least its 4-byte length prefix.
    if (std::size_t(tx_count) > (data.size() - pos) / 4) return std::unexpected(kErrInvalidBlock);

    v->txs.reserve(tx_count);
    v->tx_ids.reserve(tx_count);
    for (std::uint32_t i = 0; i < tx_count; ++i) {
        if (pos + 4 > data.size()) return std::unexpected(kErrInvalidBlock);
        const std::uint32_t tx_len = get_be32(data, pos);
        pos += 4;
        if (pos + std::size_t(tx_len) > data.size()) return std::unexpected(kErrInvalidBlock);
        auto tx = parse_transaction(data.subspan(pos, tx_len));
        if (!tx) return std::unexpected(tx.error());
        v->tx_ids.push_back(tx->id);
        v->txs.push_back(std::move(*tx));
        pos += tx_len;
    }

    if (pos != data.size()) return std::unexpected(wire::kErrTrailingBytes);

    v->set_raw(bytes_of(data));
    v->set_computed_id(v->compute_id());
    return v;
}

}  // namespace lux::zkvm

File: chains/rust/zkvm/src/vertex.rs
// SPDX-License-Identifier: BSD-3-Clause-Eco
// Copyright (C) 2019-2026, Lux Industries Inc. All rights reserved.

//! The other shape this chain decides in.
//!
//! A vertex carries transactions like a block, and differs in what it says
//! about its place: several parents, an epoch, and no timestamp. Two vertices
//! CONFLICT exactly when their nullifier sets intersect — one shielded note
//! cannot be spent by both.
//!
//! A VERTEX IS HELD TO WHAT A BLOCK IS HELD TO. It once checked its
//! transactions and nothing else — not the parents, not the height — so a
//! vertex naming no parent at height 2^40 verified, and accepting it set the
//! store's height to 2^40, pruned every block in flight, and left the linear
//! chain unable to propose a child ever again.
//!
//! The encoding here is NOT ZAP. It is the reference's own big-endian framing,
//! reproduced because these bytes are what the store writes to disk and what
//! the id is derived beside. It carries the same rule the ZAP frames do — every
//! byte handed in belongs to the vertex — for the same reason: without it,
//! trailing bytes ride along in what the store writes, one logical vertex has
//! unboundedly many encodings under one id, and a peer can park megabytes under
//! a legitimate one.

use std::collections::BTreeSet;

use crate::error::{Error, Result};
use crate::hash::Fold;
use crate::host;
use crate::ids::{self, Id};
use crate::txs::Tx;
use crate::wire;

/// A vertex's identity: the chain, its place, and what it carries.
pub fn compute_id(bind: &Id, height: u64, epoch: u32, parents: &[Id], txs: &[Tx]) -> Id {
    let mut h = Fold::new();
    h.raw(bind).num(height).num32(epoch);
    for p in parents {
        h.raw(p);
    }
    for tx in txs {
        h.raw(&tx.id());
    }
    h.id()
}

/// A DAG vertex.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Vertex {
    height: u64,
    epoch: u32,
    parents: Vec<Id>,
    txs: Vec<Tx>,
    id: Id,
    bytes: Vec<u8>,
    status: host::Status,
}

impl Vertex {
    /// A vertex this node built.
    pub fn new(bind: &Id, height: u64, epoch: u32, parents: Vec<Id>, txs: Vec<Tx>) -> Vertex {
        let id = compute_id(bind, height, epoch, &parents, &txs);
        let bytes = encode(height, epoch, &parents, &txs);
        Vertex {
            height,
            epoch,
            parents,
            txs,
            id,
            bytes,
            status: host::Status::Processing,
        }
    }

    /// A vertex off the wire.
    pub fn parse(bind: &Id, data: &[u8]) -> Result<Vertex> {
        let (height, epoch, parents, txs) = decode(data)?;
        let id = compute_id(bind, height, epoch, &parents, &txs);
        Ok(Vertex {
            height,
            epoch,
            parents,
            txs,
            id,
            bytes: data.to_vec(),
            status: host::Status::Unknown,
        })
    }

    pub fn id(&self) -> Id {
        self.id
    }

    pub fn height(&self) -> u64 {
        self.height
    }

    pub fn epoch(&self) -> u32 {
        self.epoch
    }

    pub fn parents(&self) -> &[Id] {
        &self.parents
    }

    pub fn txs(&self) -> &[Tx] {
        &self.txs
    }

    pub fn bytes(&self) -> &[u8] {
        &self.bytes
    }

    pub fn status(&self) -> host::Status {
        self.status
    }

    pub fn set_status(&mut self, s: host::Status) {
        self.status = s;
    }

    /// Everything a vertex can be refused for without reading the chain, given
    /// the frontier it claims to extend.
    ///
    /// The store keeps ONE tip for both shapes, so a vertex that names anything
    /// else moves the chain sideways.
    pub fn shape(&self, cap: u32, tip: &Id, tip_height: u64) -> Result<()> {
        if self.txs.len() as u64 > u64::from(cap) {
            return Err(Error::TxCap {
                txs: self.txs.len(),
                cap,
            });
        }
        if self.parents.len() != 1 || self.parents[0] != *tip {
            return Err(Error::NotOnTip(format!(
                "vertex parents {:?} do not name the tip {}",
                self.parents.iter().map(ids::short).collect::<Vec<_>>(),
                ids::hex(tip)
            )));
        }
        if self.height != tip_height + 1 {
            return Err(Error::InvalidHeight);
        }
        let mut here: BTreeSet<&[u8]> = BTreeSet::new();
        for tx in &self.txs {
            for n in &tx.nullifiers {
                if !here.insert(n.as_slice()) {
                    return Err(Error::DuplicateNullifier);
                }
            }
        }
        Ok(())
    }

    /// The nullifiers this vertex spends. Its conflict key.
    pub fn nullifiers(&self) -> BTreeSet<&[u8]> {
        self.txs
            .iter()
            .flat_map(|tx| tx.nullifiers.iter().map(|n| n.as_slice()))
            .collect()
    }

    /// Whether this vertex and `other` spend a note in common.
    pub fn conflicts(&self, other: &Vertex) -> bool {
        let ours = self.nullifiers();
        other.nullifiers().iter().any(|n| ours.contains(n))
    }
}

/// `height ‖ epoch ‖ parent count ‖ parents ‖ tx count ‖ (length ‖ tx)*`, all
/// big-endian.
fn encode(height: u64, epoch: u32, parents: &[Id], txs: &[Tx]) -> Vec<u8> {
    let mut out = Vec::with_capacity(8 + 4 + 4 + parents.len() * ids::ID_LEN + 4 + txs.len() * 64);
    out.extend_from_slice(&height.to_be_bytes());
    out.extend_from_slice(&epoch.to_be_bytes());
    out.extend_from_slice(&(parents.len() as u32).to_be_bytes());
    for p in parents {
        out.extend_from_slice(p);
    }
    out.extend_from_slice(&(txs.len() as u32).to_be_bytes());
    for tx in txs {
        let raw = wire::write_tx(tx);
        out.extend_from_slice(&(raw.len() as u32).to_be_bytes());
        out.extend_from_slice(&raw);
    }
    out
}

/// The counts are attacker-controlled, so each is bounded by the bytes that
/// remain BEFORE anything is allocated: a 16-byte vertex claiming 2^32-1
/// parents would otherwise ask for 128 GiB and take the node with it.
#[allow(clippy::type_complexity)]
fn decode(data: &[u8]) -> Result<(u64, u32, Vec<Id>, Vec<Tx>)> {
    if data.len() < 16 {
        return Err(Error::InvalidBlock);
    }
    let mut pos = 0usize;
    let take = |pos: &mut usize, n: usize| -> &[u8] {
        let s = &data[*pos..*pos + n];
        *pos += n;
        s
    };

    let height = u64::from_be_bytes(take(&mut pos, 8).try_into().unwrap());
    let epoch = u32::from_be_bytes(take(&mut pos, 4).try_into().unwrap());

    let parent_count = u32::from_be_bytes(take(&mut pos, 4).try_into().unwrap()) as usize;
    if parent_count > (data.len() - pos) / ids::ID_LEN {
        return Err(Error::InvalidBlock);
    }
    let mut parents = Vec::with_capacity(parent_count);
    for _ in 0..parent_count {
        parents.push(ids::from_slice(take(&mut pos, ids::ID_LEN)));
    }

    if pos + 4 > data.len() {
        return Err(Error::InvalidBlock);
    }
    let tx_count = u32::from_be_bytes(take(&mut pos, 4).try_into().unwrap()) as usize;
    // Every transaction costs at least its four-byte length prefix.
    if tx_count > (data.len() - pos) / 4 {
        return Err(Error::InvalidBlock);
    }
    let mut txs = Vec::with_capacity(tx_count);
    for _ in 0..tx_count {
        if pos + 4 > data.len() {
            return Err(Error::InvalidBlock);
        }
        let len = u32::from_be_bytes(take(&mut pos, 4).try_into().unwrap()) as usize;
        if pos + len > data.len() {
            return Err(Error::InvalidBlock);
        }
        txs.push(wire::read_tx(take(&mut pos, len))?);
    }

    if pos != data.len() {
        return Err(Error::TrailingBytes);
    }
    Ok((height, epoch, parents, txs))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::block::binding;
    use crate::txs::{Kind, Proof, Shielded, Tx};

    fn bind() -> Id {
        binding(&ids::repeated(40), 1)
    }

    fn transfer(nullifier: u8) -> Tx {
        Tx {
            kind: Kind::TRANSFER,
            version: 1,
            nullifiers: vec![vec![nullifier; 32]],
            outputs: vec![Shielded {
                commitment: vec![0x50; 32],
                ..Shielded::default()
            }],
            proof: Some(Proof {
                system: b"stark".to_vec(),
                data: vec![0x60; 192],
                public: vec![vec![0x61; 32]],
            }),
            fee: 1,
            expiry: 1000,
            ..Tx::default()
        }
    }

    fn vertex(height: u64, parent: Id, txs: Vec<Tx>) -> Vertex {
        Vertex::new(&bind(), height, 0, vec![parent], txs)
    }

    #[test]
    fn a_vertex_round_trips_through_its_own_framing() {
        for txs in [vec![], vec![transfer(1)], vec![transfer(1), transfer(2)]] {
            let v = vertex(1, ids::repeated(9), txs);
            let back = Vertex::parse(&bind(), v.bytes()).unwrap();
            assert_eq!(back.id(), v.id());
            assert_eq!(back.height(), v.height());
            assert_eq!(back.epoch(), v.epoch());
            assert_eq!(back.parents(), v.parents());
            assert_eq!(back.txs(), v.txs());
        }
    }

    #[test]
    fn every_field_is_in_the_identity() {
        let base = compute_id(&bind(), 1, 0, &[ids::repeated(9)], &[transfer(1)]);
        assert_ne!(base, compute_id(&bind(), 2, 0, &[ids::repeated(9)], &[transfer(1)]));
        assert_ne!(base, compute_id(&bind(), 1, 1, &[ids::repeated(9)], &[transfer(1)]));
        assert_ne!(base, compute_id(&bind(), 1, 0, &[ids::repeated(8)], &[transfer(1)]));
        assert_ne!(base, compute_id(&bind(), 1, 0, &[ids::repeated(9)], &[transfer(2)]));
        assert_ne!(
            base,
            compute_id(&binding(&ids::repeated(41), 1), 1, 0, &[ids::repeated(9)], &[transfer(1)])
        );
    }

    #[test]
    fn bytes_past_the_end_are_refused_rather_than_carried_along() {
        let v = vertex(1, ids::repeated(9), vec![transfer(1)]);
        let mut raw = v.bytes().to_vec();
        raw.push(0xFF);
        assert_eq!(Vertex::parse(&bind(), &raw), Err(Error::TrailingBytes));
    }

    #[test]
    fn a_truncated_vertex_is_refused_at_every_cut() {
        let v = vertex(1, ids::repeated(9), vec![transfer(1)]);
        let raw = v.bytes();
        for cut in [0, 8, 15, 16, raw.len() / 2, raw.len() - 1] {
            assert!(Vertex::parse(&bind(), &raw[..cut]).is_err(), "cut at {cut}");
        }
    }

    /// The count is bounded before anything is allocated.
    #[test]
    fn an_absurd_parent_count_allocates_nothing() {
        let mut raw = Vec::new();
        raw.extend_from_slice(&1u64.to_be_bytes());
        raw.extend_from_slice(&0u32.to_be_bytes());
        raw.extend_from_slice(&u32::MAX.to_be_bytes());
        assert_eq!(Vertex::parse(&bind(), &raw), Err(Error::InvalidBlock));
    }

    #[test]
    fn an_absurd_transaction_count_allocates_nothing() {
        let mut raw = Vec::new();
        raw.extend_from_slice(&1u64.to_be_bytes());
        raw.extend_from_slice(&0u32.to_be_bytes());
        raw.extend_from_slice(&0u32.to_be_bytes());
        raw.extend_from_slice(&u32::MAX.to_be_bytes());
        assert_eq!(Vertex::parse(&bind(), &raw), Err(Error::InvalidBlock));
    }

    /// The rule the vertex once did not have. A vertex naming no parent, at a
    /// height nothing reached, is refused.
    #[test]
    fn a_vertex_that_names_no_parent_is_refused() {
        let v = Vertex::new(&bind(), 1 << 40, 0, vec![], vec![]);
        assert!(matches!(
            v.shape(100, &ids::repeated(9), 0),
            Err(Error::NotOnTip(_))
        ));
    }

    #[test]
    fn a_vertex_extends_the_tip_and_the_height_above_it() {
        let tip = ids::repeated(9);
        assert_eq!(vertex(1, tip, vec![]).shape(100, &tip, 0), Ok(()));
        assert_eq!(
            vertex(5, tip, vec![]).shape(100, &tip, 0),
            Err(Error::InvalidHeight)
        );
        assert!(matches!(
            vertex(1, ids::repeated(3), vec![]).shape(100, &tip, 0),
            Err(Error::NotOnTip(_))
        ));
    }

    #[test]
    fn a_vertex_over_the_cap_is_refused() {
        let tip = ids::repeated(9);
        let txs: Vec<Tx> = (0..3u8).map(transfer).collect();
        assert_eq!(
            vertex(1, tip, txs).shape(2, &tip, 0),
            Err(Error::TxCap { txs: 3, cap: 2 })
        );
    }

    #[test]
    fn one_note_spent_twice_inside_one_vertex_is_refused() {
        let tip = ids::repeated(9);
        let mut twin = transfer(1);
        twin.fee = 2;
        assert_eq!(
            vertex(1, tip, vec![transfer(1), twin]).shape(100, &tip, 0),
            Err(Error::DuplicateNullifier)
        );
    }

    #[test]
    fn two_vertices_conflict_exactly_when_they_spend_a_note_in_common() {
        let tip = ids::repeated(9);
        let a = vertex(1, tip, vec![transfer(1), transfer(2)]);
        let shares = vertex(1, tip, vec![transfer(2), transfer(3)]);
        let disjoint = vertex(1, tip, vec![transfer(3)]);
        assert!(a.conflicts(&shares));
        assert!(shares.conflicts(&a));
        assert!(!a.conflicts(&disjoint));
        assert!(!vertex(1, tip, vec![]).conflicts(&a));
    }
}
